### PR TITLE
jupyter_ext_dev: pull request for jupyter extension 2.x and 3.x recap…

### PR DESCRIPTION
**Pull request by Johnsonms**
… #4141

recapped: https://github.com/microsoft/nni/pull/4141/commits

### CODE REVIEW

File: nni/tools/jupyter_extension/management.py [hunk 1]
```diff
@@ -1,11 +1,15 @@
 import json
 from pathlib import Path
 import shutil
+import os
```
Comment (cruiseliu) [inline nni/tools/jupyter_extension/management.py:4]: This is never used in this version.

File: setup.py [hunk 2]
```diff
@@ -58,13 +62,25 @@
 import shutil
 import sys
 
+import jupyterlab
```
Comment (cruiseliu) [inline setup.py:65]: Please add try-catch for this import. The script should not crash when building machine has not installed jupyter.
And I personally suggest put this and `jupyter_lab_version = ...` into a function `get_jupyter_lab_version()`.

File: setup_ts.py [hunk 3]
```diff
@@ -22,10 +22,11 @@
 import traceback
 from zipfile import ZipFile
 
+import jupyterlab
```
Comment (cruiseliu) [inline setup_ts.py:25]: Same as `setup.py`.

File: setup_ts.py [hunk 4]
```diff
@@ -177,7 +198,10 @@ def symlink_nni_node():
 
     _symlink('ts/webui/build', 'nni_node/static')
 
-    if Path('ts/jupyter_extension/dist').exists():
+    if jupyter_lab_version == '2':
+        _symlink('ts/jupyter_extension/build', 'nni_node/jupyter-extension')
+        _symlink(sys.exec_prefix+'/share/jupyter/lab/extensions', 'nni_node/jupyter-extension/extensions')
```
Comment (ultmaster) [inline setup_ts.py:203]: Using `+` to glue two paths is not recommended.

File: setup_ts.py [hunk 5]
```diff
@@ -120,6 +123,24 @@ def download_toolchain():
...
+        package_json['dependencies']['@jupyterlab/application'] = '^2.3.0'
+        package_json['dependencies']['@jupyterlab/launcher'] = '^2.3.0'
+
+        package_json['jupyterlab']['outputDir'] = 'build'
+        json.dump(package_json, open('ts/jupyter_extension/package.json', 'w'), indent=2)
+        print(f'updated package.json with {json.dumps(package_json, indent=2)}')
+
+def restore_package():
+    if jupyter_lab_version == '2':
+        package_json = json.load(open('ts/jupyter_extension/.package_default.json'))
```
Comment (ultmaster) [inline setup_ts.py:141]: Delete `.package_default.json` or put it into gitignore.

File: nni/tools/jupyter_extension/management.py (hunk 1 above)
Comment (Johnsonms) [reply nni/tools/jupyter_extension/management.py:4]: L 44 or 45, using os.path or os.symlink

File: setup.py (hunk 2 above)
Comment (Johnsonms) [reply setup.py:65]: changed, In this way, the building machine without jupyterlab installed will report error and the package related jupyter lab extension will not built, but continue to build other package

File: setup_ts.py (hunk 3 above)
Comment (Johnsonms) [reply setup_ts.py:25]: changed

File: setup_ts.py (hunk 5 above)
Comment (Johnsonms) [reply setup_ts.py:141]: done, make sense

**Review by ultmaster — APPROVED** [review]

File: setup.py [hunk 6]
```diff
@@ -65,6 +69,26 @@
...
+def check_jupyter_lab_version():
+    environ_version = os.environ.get('JUPYTER_LAB_VERSION')
+
+    jupyter_lab_version = _get_jupyter_lab_version()
+
+    if environ_version:
+        if jupyter_lab_version.split('.')[0] != environ_version.split('.')[0]:
+            sys.exit(f'ERROR: To build a jupyter lab extension, run "JUPYTER_LAB_VERSION={jupyter_lab_version}", current: {environ_version} ')
+    elif jupyter_lab_version.split('.')[0] != '3':
+            sys.exit(f'ERROR: To build a jupyter lab extension, run "JUPYTER_LAB_VERSION={jupyter_lab_version}" first for nondefault version(3.x)')
```
Comment (liuzhe-lz) [inline setup.py:90]: Bad indent

Comment (Johnsonms) [reply setup.py]: fixed it

**Review by liuzhe-lz — APPROVED** [review]

**Pull request merged**